tools/lambda-promtail/lambda-promtail/main.go

```diff
@@ -63,71 +54,20 @@ func init() {
 	fmt.Println("keep stream: ", keepStream)
 }
 
-func handler(ctx context.Context, ev events.CloudwatchLogsEvent) error {
-	data, err := ev.AWSLogs.Parse()
-	if err != nil {
-		fmt.Println("error parsing log event: ", err)
-		return err
-	}
-	labels := model.LabelSet{
-		model.LabelName("__aws_cloudwatch_log_group"): model.LabelValue(data.LogGroup),
-		model.LabelName("__aws_cloudwatch_owner"):     model.LabelValue(data.Owner),
-	}
-	if keepStream {
-		labels[model.LabelName("__aws_cloudwatch_log_stream")] = model.LabelValue(data.LogStream)
-	}
-
-	stream := logproto.Stream{
-		Labels:  labels.String(),
-		Entries: make([]logproto.Entry, 0, len(data.LogEvents)),
+func handler(ctx context.Context, ev map[string]interface{}) error {
+	s3event, err := checkIfS3Event(ev)
```

Comment:
sorry, I don't understand what you meant with `w.(type)`, do you mean `ev.(type)`? if thats the case then it won't work, it will just return `map[string]interface{}`

the switch statement makes sense

Reply:
Hmm it seems like what you're actually doing might be the cleanest way since the event would need to be json umarshal'd in order to check the golang type via type assertion. Another option would be to use the json field from the s3 event that tells us the event source `"eventSource":"aws:s3",` see [here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-content-structure.html).

Personally what I would do is have one function `convertToEvent` that returns the decoded event or an error if it's not an event type we have a `checkIfTEvent` function for. `handler` can then `switch ev.(type)` and have case statements for the actual `events.S3Event` or `events.CloudwatchLogsEvent` cases. What do you think?

Reply:
> Another option would be to use the json field from the s3 event that tells us the event source `"eventSource":"aws:s3"`
This only works for s3 and some other events, not all of them have this field (e.g. CloudWatch) and I wanted to keep it flexible enough to add other event types

> Personally what I would do is have one function `convertToEvent` that returns the decoded event
sure but how do I return different event types in Go? Generics (haven't look that yet)?

Reply:
Generics aren't available until 1.18, which is still in beta AFAICT. Should be able to do it even without generics, `convertToEvent` can marshal the json data from the event into the correct struct type and return`interface{}` and then we can have a switch statement on the type assertion. It would look something like this:
```
func checkEventType(a interface{}) {
	switch a.(type) {
	case events.S3Event:
		fmt.Println("object is an s3 event")

	case events.CloudwatchLogsEvent:
		fmt.Println("object is a CWL event")
	default:
		fmt.Println("unknown event type")
	}
}

func main() {
	var s3event events.S3Event
	var cwlEvent events.CloudwatchLogsEvent

	checkEventType(s3event)
	checkEventType(cwlEvent)
}
```

Reply:
@cstyan please review the new changes, I think I got what you meant